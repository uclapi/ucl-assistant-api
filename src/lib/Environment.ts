const {
  UCLAPI_TOKEN,
  UCLAPI_CLIENT_ID,
  UCLAPI_CLIENT_SECRET,
  SECRET,
  API_URL,
  NOTIFICATIONS_URL,
  REDIS_URL,
  NODE_ENV,
  TEST_MODE,
  PORT,
} = process.env

class Environment {
  static TOKEN = UCLAPI_TOKEN
  static CLIENT_ID = UCLAPI_CLIENT_ID
  static CLIENT_SECRET = UCLAPI_CLIENT_SECRET

  static SECRET = SECRET

  static API_URL = API_URL
  static NOTIFICATIONS_URL = NOTIFICATIONS_URL
  static REDIS_URL = REDIS_URL

  static NODE_ENV = NODE_ENV
  static DEVELOPMENT_MODE = NODE_ENV === `development`
  static PRODUCTION_MODE = NODE_ENV === `production`
  static TEST_MODE = TEST_MODE === `true`
  static PORT = PORT
}

export default Environment