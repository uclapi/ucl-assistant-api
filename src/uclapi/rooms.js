const {
  ROOMS_SEARCH_URL,
  ROOMS_EQUIPMENT_URL,
} = require(`../constants/apiRoutes`)
const axios = require(`axios`)

const { UCLAPI_TOKEN } = process.env

const roomsSearch = async query => {
  if (!query || query.length <= 3) {
    throw new Error(`Must provide a query!`)
  }

  return (await axios.get(ROOMS_SEARCH_URL, {
    params: {
      token: UCLAPI_TOKEN,
      roomname: query,
    },
  })).data
}

const getSites = async () => {
  const { data: { rooms } } = (await axios.get(ROOMS_SEARCH_URL, {
    params: {
      token: UCLAPI_TOKEN,
    },
  }))
  const sites = Array.from(new Set(rooms.map(({ siteid }) => siteid)))
    .map(siteid => {
      return {
        siteid,
        sitename: rooms.find(r => r.siteid === siteid).sitename,
      }
    })
  return { sites }
}

const getEquipment = async (roomid, siteid) => {
  if (!roomid || !siteid) {
    throw new Error(`Must specify roomid and siteid`)
  }

  return (await axios.get(ROOMS_EQUIPMENT_URL, {
    params: {
      token: UCLAPI_TOKEN,
      roomid,
      siteid,
    },
  })).data
}

module.exports = {
  roomsSearch,
  getEquipment,
  getSites,
}
