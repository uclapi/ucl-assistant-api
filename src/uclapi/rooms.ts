import axios from 'axios'
import ApiRoutes from '../constants/apiRoutes'

const { UCLAPI_TOKEN } = process.env

export const roomsSearch = async (query: string) => {
  if (!query || query.length <= 3) {
    throw new Error(`Must provide a query!`)
  }

  return (await axios.get(ApiRoutes.ROOMS_SEARCH_URL, {
    params: {
      token: UCLAPI_TOKEN,
      roomname: query,
    },
  })).data
}

export const getSites = async () => {
  const { data: { rooms } } = (await axios.get(ApiRoutes.ROOMS_SEARCH_URL, {
    params: {
      token: UCLAPI_TOKEN,
    },
  }))
  const sites = Array.from(new Set(rooms.map(({ siteid }) => siteid)))
    .map(siteid => {
      return {
        siteid,
        sitename: rooms.find(r => r.siteid === siteid).sitename,
      }
    })
  return { sites }
}

export const getEquipment = async (roomid, siteid) => {
  if (!roomid || !siteid) {
    throw new Error(`Must specify roomid and siteid`)
  }

  return (await axios.get(ApiRoutes.ROOMS_EQUIPMENT_URL, {
    params: {
      token: UCLAPI_TOKEN,
      roomid,
      siteid,
    },
  })).data
}
