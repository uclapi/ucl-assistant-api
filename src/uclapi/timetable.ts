import axios from 'axios'
import moment from 'moment'
import ApiRoutes from '../constants/apiRoutes'
import ErrorManager from '../lib/ErrorManager'

export const getPersonalWeekTimetable = async (token: string, date = null) => {
  if (!date) {
    throw new Error(`Must specify date to retrieve weekly timetable`)
  }
  const momentDate = moment(date)
  if (momentDate.day() !== 1) {
    throw new Error(`Date must be a Monday`)
  }

  const params = {
    client_secret: process.env.UCLAPI_CLIENT_SECRET,
    token,
  }

  const weekdays = [...new Array(5)].map(
    (_, index) => momentDate.clone().add(index, `days`).format(`YYYY-MM-DD`),
  )

  return {
    lastModified: (new Date()).toUTCString(),
    data: {
      timetable: (await Promise.all(
        weekdays.map(
          date => axios.get(
            ApiRoutes.PERSONAL_TIMETABLE_URL,
            {
              params: {
                ...params,
                'date_filter': date,
              },
            },
          ),
        ),
      )).map(({ data: { timetable } }) => timetable)
        .reduce((prev, cur) => {
          const [key, value] = Object.entries(cur)[0]
          return {
            ...prev,
            [key]: value,
          }
        }, {}),
    },
  }
}

export const getPersonalTimetable = async (token, date = null) => {
  const params = {
    client_secret: process.env.UCLAPI_CLIENT_SECRET,
    token,
  }

  if (date) {
    params[`date_filter`] = moment(date).format(`YYYY-MM-DD`)
  }
  try {
    const { data } = await axios.get(
      ApiRoutes.PERSONAL_TIMETABLE_URL,
      { params },
    )

    return {
      // lastModified should be headers[`last-modified`]
      // but the main API is returning the wrong value
      // TODO: change this back when caching is fixed
      lastModified: (new Date()).toUTCString(),
      data,
    }
  } catch (error) {
    ErrorManager.captureError(error)
    throw error
  }

}

export const getModuleTimetable = async (token, timetableModule) => {
  const { data, headers } = await axios.get(ApiRoutes.MODULE_TIMETABLE_URL, {
    params: {
      client_secret: process.env.UCLAPI_CLIENT_SECRET,
      token,
      modules: timetableModule,
    },
  })
  return {
    lastModified: headers[`last-modified`],
    data,
  }
}
