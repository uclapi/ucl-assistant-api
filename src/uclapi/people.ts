import axios from 'axios'
import ApiRoutes from '../constants/apiRoutes'
import Environment from '../lib/Environment'
import ErrorManager from '../lib/ErrorManager'

export const peopleSearch = async (query: string) => {
  if (!query || query.length <= 3) {
    throw new Error(`Must provide a query!`)
  }

  try {
    const { data } = await axios.get(ApiRoutes.PEOPLE_SEARCH_URL, {
      params: {
        token: Environment.TOKEN,
        client_secret: Environment.CLIENT_SECRET,
        query,
      },
    })

    return data
  } catch (error) {
    ErrorManager.captureError(error)
    return {
      ok: false,
      error,
      people: [],
    }
  }
  
}
