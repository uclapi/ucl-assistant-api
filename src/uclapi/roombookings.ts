import axios from 'axios'
import moment from 'moment'
import ApiRoutes from '../constants/apiRoutes'
import Environment from '../lib/Environment'

const { UCLAPI_TOKEN } = process.env

export const getRoomBookings = async ({ roomid, siteid, date }) => {
  if (!roomid) {
    throw new Error(`Must provide a roomid!`)
  }
  if (!siteid) {
    throw new Error(`Must provide a siteid`)
  }

  const { data } = await axios.get(ApiRoutes.ROOMBOOKINGS_DATA_URL, {
    params: {
      token: Environment.TOKEN,
      roomid,
      siteid,
      date,
    },
  })
  return data
}

export const getFreeRooms = async (
  startDateTime = new Date().toISOString(),
  endDateTime = moment()
    .endOf(`day`)
    .toISOString(),
) => {

  const { data } =  await axios.get(ApiRoutes.ROOMBOOKINGS_FREEROOMS_URL, {
    params: {
      token: Environment.TOKEN,
      start_datetime: startDateTime,
      end_datetime: endDateTime,
    },
  })
  return data
}

export default {
  getRoomBookings,
  getFreeRooms,
}