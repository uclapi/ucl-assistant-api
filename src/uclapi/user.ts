import axios from 'axios'
import ApiRoutes from '../constants/apiRoutes'
import Environment from '../lib/Environment'

export const getToken = async (code: string) => {
  const { data } = await axios.get(
    ApiRoutes.USER_TOKEN_URL, {
    params: {
      client_id: Environment.CLIENT_ID,
      client_secret: Environment.CLIENT_SECRET,
      code,
    },
  })
  return data
}

export const getUserData = async (token: string) => {
  const { data } = await axios.get(
    ApiRoutes.USER_DATA_URL, {
    params: {
      client_secret: Environment.CLIENT_SECRET,
      token,
    },
  })
  return data
}

export default {
  getToken,
  getUserData,
}
