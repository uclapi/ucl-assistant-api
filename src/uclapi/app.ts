import Koa from 'koa'
import Router from 'koa-router'
import { jwt } from '../middleware/auth'
import redis from '../redis'
import { peopleSearch } from './people'
import { getFreeRooms, getRoomBookings } from './roombookings'
import { getEquipment, getSites, roomsSearch } from './rooms'
import {
  getModuleTimetable,
  getPersonalTimetable,
  getPersonalWeekTimetable,
} from './timetable'
import { getUserData } from './user'
import Workspaces from './workspaces'

const app = new Koa()
const router = new Router()

router.get(`/user`, jwt, async ctx => {
  ctx.body = await getUserData(ctx.state.user.apiToken)
})

router.get(`/timetable`, jwt, async ctx => {
  const date = ctx.query.date || null

  const timetableData = await redis.loadOrFetch(
    ctx,
    `${redis.keys.TIMETABLE_PERSONAL_PATH}/${ctx.state.user.upi}/${date}`,
    async () => getPersonalTimetable(ctx.state.user.apiToken, date),
    redis.ttl.TIMETABLE_TTL,
  )

  const { lastModified, data } = timetableData
  ctx.body = data
  ctx.set(`Last-Modified`, lastModified)
})

router.get(`/timetable/week`, jwt, async ctx => {
  const date = ctx.query.date || null

  const timetableData = await redis.loadOrFetch(
    ctx,
    `${redis.keys.TIMETABLE_PERSONAL_PATH}/${ctx.state.user.upi}/week/${date}`,
    async () => getPersonalWeekTimetable(ctx.state.user.apiToken, date),
    redis.ttl.TIMETABLE_TTL,
  )

  const { lastModified, data } = timetableData
  ctx.body = data
  ctx.set(`Last-Modified`, lastModified)
})

router.get(`/timetable/:module`, jwt, async ctx => {
  ctx.assert(ctx.params.module, 400)
  const { module: timetableModule } = ctx.params
  const date = ctx.query.date || null

  const timetableData = await redis.loadOrFetch(
    ctx,
    `${redis.keys.TIMETABLE_MODULE_PATH}/${timetableModule}/${date}`,
    async () => getModuleTimetable(
      ctx.state.user.apiToken,
      timetableModule,
    ),
    redis.ttl.TIMETABLE_TTL,
  )

  const { lastModified, data } = timetableData
  ctx.body = data
  ctx.set(`Last-Modified`, lastModified)
})

router.get(`/search/people`, jwt, async ctx => {
  ctx.assert(
    (ctx.query.query || ``).length >= 3,
    400,
    `Query must be at least three characters long`,
  )
  const data = await redis.loadOrFetch(
    ctx,
    `${redis.keys.PEOPLE_SEARCH_PATH}/${ctx.query.query}`,
    async () => peopleSearch(ctx.query.query),
    redis.ttl.PEOPLE_SEARCH_TTL,
  )
  ctx.body = data
})

router.get(`/search/rooms`, jwt, async ctx => {
  ctx.assert(
    (ctx.query.query || ``).length >= 3,
    400,
    `Query must be at least four characters long`,
  )
  const data = await redis.loadOrFetch(
    ctx,
    `${redis.keys.ROOMS_SEARCH_PATH}/${ctx.query.query}`,
    async () => roomsSearch(ctx.query.query),
    redis.ttl.ROOMS_SEARCH_TTL,
  )
  ctx.body = data
})

router.get(`/sites`, jwt, async ctx => {
  const rooms = await redis.loadOrFetch(
    ctx,
    redis.keys.SITES_SEARCH_PATH,
    async () => getSites(),
    redis.ttl.ROOMS_SEARCH_TTL,
  )
  ctx.body = rooms
})

router.get(`/equipment`, jwt, async ctx => {
  ctx.assert(ctx.query.roomid, 400, `Must specify roomid`)
  ctx.assert(ctx.query.siteid, 400, `Must specify siteid`)
  const data = await redis.loadOrFetch(
    ctx,
    `${
      redis.keys.WORKSPACE_EQUIPMENT_PATH
    }/${ctx.query.roomid}/${ctx.query.siteid}`,
    async () => getEquipment(ctx.query.roomid, ctx.query.siteid),
    redis.ttl.WORKSPACE_EQUIPMENT_TTL,
  )
  ctx.body = data
})

router.get(`/workspaces/getimage/:id.png`, jwt, async ctx => {
  ctx.assert(ctx.params.id, 400, `Must specify id`)
  ctx.set({ "Content-Type": `image/png` })
  ctx.state.jsonify = false
  const res = await Workspaces.getImage(ctx.params.id)
  ctx.body = res
})

router.get(`/workspaces/getliveimage/map.svg`, jwt, async ctx => {
  ctx.assert(ctx.query.survey_id)
  ctx.assert(ctx.query.map_id)
  ctx.set({ "Content-Type": `image/svg+xml` })
  ctx.state.jsonify = false
  const res = await Workspaces.getLiveImage({
    surveyId: ctx.query.survey_id,
    mapId: ctx.query.map_id,
    circleRadius: ctx.query.circle_radius,
    absentColour: ctx.query.absent_colour,
    occupiedColour: ctx.query.occupied_colour,
  })
  ctx.body = res
})

router.get(`/workspaces/summary`, jwt, async ctx => {
  const { data, lastModified } = await redis.loadOrFetch(
    ctx,
    redis.keys.WORKSPACE_SUMMARY_PATH,
    async () => Workspaces.getAllSeatInfo(),
    redis.ttl.WORKSPACE_SUMMARY_TTL,
  )
  ctx.body = data
  ctx.set(`Last-Modified`, lastModified)
})

router.get(`/workspaces/historic`, jwt, async ctx => {
  ctx.assert(ctx.query.id, 400, `Need to include a survey id.`)
  const data = await redis.loadOrFetch(
    ctx,
    `${redis.keys.WORKSPACE_HISTORIC_DATA_PATH}/${ctx.query.id}`,
    async () => Workspaces.getHistoricSeatInfo(ctx.query.id),
    redis.ttl.WORKSPACE_HISTORIC_DATA_TTL,
  )
  ctx.body = data
})

router.get(`/workspaces/:id/seatinfo`, jwt, async ctx => {
  ctx.assert(ctx.params.id, 400)
  ctx.body = await Workspaces.getSeatingInfo(ctx.params.id)
})

router.get(`/workspaces`, jwt, async ctx => {
  const surveyFilter = ctx.query.survey_filter
    ? ctx.query.survey_filter
    : `student`
  ctx.body = await redis.loadOrFetch(
    ctx,
    `${redis.keys.WORKSPACE_SURVEYS_PATH}/${surveyFilter}`,
    async () => Workspaces.getWorkspaces(surveyFilter),
    redis.ttl.WORKSPACE_SURVEYS_TTL,
  )
})

router.get(`/roombookings`, jwt, async ctx => {
  ctx.assert(ctx.query.roomid, 400, `Please include a roomid`)
  ctx.assert(ctx.query.siteid, 400, `Please include a siteid`)
  ctx.assert(ctx.query.date, 400, `Please include a date`)
  ctx.body = await getRoomBookings({
    roomid: ctx.query.roomid,
    siteid: ctx.query.siteid,
    date: ctx.query.date,
  })
})

router.get(`/freerooms`, jwt, async ctx => {
  const {
    start_datetime: startDateTime,
    end_datetime: endDateTime,
  } = ctx.query
  ctx.body = await getFreeRooms(startDateTime, endDateTime)
})

app.use(router.routes())
app.use(router.allowedMethods())

export default app
