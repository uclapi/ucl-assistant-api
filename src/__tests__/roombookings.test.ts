import axios from 'axios'
import Environment from '../lib/Environment'
import { getRoomBookings, getFreeRooms } from '../uclapi/roombookings'

jest.mock(`axios`)
const axiosGet = axios.get as jest.Mock

const SAMPLE_TOKEN = `tokeytoken`
const SAMPLE_SECRET = `shibboleth`

describe(`roombookings`, () => {

  beforeAll(() => {
    Environment.TOKEN = SAMPLE_TOKEN
    Environment.CLIENT_SECRET = SAMPLE_SECRET
  })

   beforeEach(() => {
    axiosGet.mockClear()
  })

  it(`should send a valid room bookings search request`, () => {
    axiosGet.mockImplementation(() => Promise.resolve({
      data: {
        ok: true,
        bookings: [],
      },
    }))
    getRoomBookings({ roomid: `123`,
      siteid: `456`,
      date: `2021-05-27T04:55:01.694Z`,
    })
    expect(axiosGet.mock.calls).toMatchSnapshot()
  })

  it(`should send a valid free rooms request`, () => {
    axiosGet.mockImplementation(() => Promise.resolve({
      data: {
        ok: true,
        free_rooms: [],
      },
    }))
    getFreeRooms(
      `2021-05-27T04:55:57.727Z`,
      `2021-05-27T04:56:50.191Z`,
    )
    expect(axiosGet.mock.calls).toMatchSnapshot()
  })
})
