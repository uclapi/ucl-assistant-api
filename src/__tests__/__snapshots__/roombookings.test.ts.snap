// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`roombookings should send a valid free rooms request 1`] = `
Array [
  Array [
    "https://uclapi.com/roombookings/freerooms",
    Object {
      "params": Object {
        "end_datetime": "2021-05-27T04:56:50.191Z",
        "start_datetime": "2021-05-27T04:55:57.727Z",
        "token": "tokeytoken",
      },
    },
  ],
]
`;

exports[`roombookings should send a valid room bookings search request 1`] = `
Array [
  Array [
    "https://uclapi.com/roombookings/bookings",
    Object {
      "params": Object {
        "date": "2021-05-27T04:55:01.694Z",
        "roomid": "123",
        "siteid": "456",
        "token": "tokeytoken",
      },
    },
  ],
]
`;
